aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+bool NoneType::isSubtypeOfExt(const TypePtr& rhs, std::ostream *why_not) const {
+  if (rhs->kind() == OptionalType::Kind) {
+    return true;
+  }
+  if (auto rhs_ = rhs->cast<UnionType>()) {
+    return rhs_->can_hold_none();
+  }
+  return Type::isSubtypeOfExt(rhs, why_not);
+}
+
+// Remove nested Optionals/Unions during the instantiation of a Union or
+// an Optional. This populates `types` with all the types found during
+// flattening
+void flattenUnion(TypePtr& type, std::vector<TypePtr>& types) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flattenUnion(inner_type, types);
+    }
+  } else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind
+        || inner_type->kind() == OptionalType::Kind) {
+      flattenUnion(inner_type, types);
+    } else {
+      types.emplace_back(inner_type);
+    }
+    types.emplace_back(NoneType::get());
+  } else {
+    types.emplace_back(type);
+  }
+}
+
+// Custom comparison for Union type auxiliary functions. We compare on
+// Kind (fast), then compare by the serialized representation
+// (slower, but guaranteed to break ties). We need to have the
+// TypeEqual struct for use in a `std::unordered_map` later.
+struct TypeEqual {
+  bool operator()(const TypePtr& a, const TypePtr& b) const {
+    return a->kind() == b->kind() && a->str() == b->str();;
+  }
+};
+
+// Helper function for `standardizeUnion`
+void filterDuplicateSubtypes(std::vector<TypePtr>& types) {
+  // If we call `unifyTypes(T, None)`, then `unifyTypes` returns
+  // `Optional[T]`. The `Optional` constructor calls the `Union`
+  // constructor, which calls `filterDuplicateSubtypes`...infinite
+  // recursion (until the stack explodes). So, let's just wrap
+  // `unifyTypes` to ensure that this condition doesn't happen. We
+  // only need to worry about this in the constructor
+  auto wrapper = [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+    // If we would have unified the types to an Optional
+    if ((t1 == NoneType::get()) ^ (t2 == NoneType::get())) {
+      return c10::nullopt;
+    }
+    return unifyTypes(t1, t2);
+  };
+
+  // Coalesce types and delete all duplicates. Moving from right to left
+  // through the vector, we try to unify the current element (`i`) with
+  // each element (`j`) before the "new" end of the vector (`end`).
+  // If we're able to unify the types at `types[i]` and `types[j]`, we
```

Comment:
this is describing the code as it exists, could you try to explain the motivation behind it?